File: pycab/cap/conn.py
import asyncio
import nats
from nats.aio.client import Client, Subscription
from . import object

class Conn:
    def __init__(self, addr: str):
        self.addr = addr
        self.nc: Client = None
        self.objs: dict[str, object.Object] = {}

    async def connect(self):
        self.nc = await nats.connect(self.addr)

    def connected_uri(self):
        return self.addr

    async def close(self):
        await self.nc.close()

    def new_object(self, objectId):
        o = object.Object(self.nc, objectId)
        self.objs[objectId] = o
    
    def get_object_ids(self):
        return self.objs.keys()
    
    def get_object(self, objectId):
        return self.objs[objectId]
    
    def register_object(self, objectId):
        return self.ensure_object(objectId)
    
    def unregister_object(self, objectId):
        if objectId in self.objs:
            del self.objs[objectId]
    
    def ensure_object(self, objectId):
        if objectId not in self.objs:
            self.objs[objectId] = object.Object(self.nc, objectId)            
        return self.objs[objectId]

    

File: gocab/cab/conn.go
package olink

import (
	"sync"

	"github.com/nats-io/nats.go"
	"github.com/rs/zerolog/log"
)

type conn struct {
	lock    sync.RWMutex
	nc      *nats.Conn
	objects map[string]*Object
}

func NewConn(addr string) (*conn, error) {
	nc, err := nats.Connect(addr)
	if err != nil {
		log.Error().Err(err).Msg("failed to connect to NATS server")
		return nil, err
	}
	return &conn{
		lock:    sync.RWMutex{},
		nc:      nc,
		objects: make(map[string]*Object),
	}, nil
}

func (c *conn) Close() {
	for _, o := range c.objects {
		o.Unsubscribe()
	}
	c.nc.Close()
}

func (c *conn) ConnectedUrl() string {
	return c.nc.ConnectedUrl()
}

func (c *conn) NewObject(id string) *Object {
	c.lock.Lock()
	defer c.lock.Unlock()
	if o, ok := c.objects[id]; ok {
		return o
	}
	o := NewObject(c.nc, id)
	c.objects[id] = o
	return o
}

// --------------------------------------------
// Registry
// --------------------------------------------

// ObjectIds returns the object ids.
func (c *conn) ObjectIds() []string {
	c.lock.RLock()
	defer c.lock.RUnlock()
	ids := make([]string, 0, len(c.objects))
	for id := range c.objects {
		ids = append(ids, id)
	}
	return ids
}

func (c *conn) Object(objectId string) *Object {
	c.lock.RLock()
	defer c.lock.RUnlock()
	return c.objects[objectId]
}

func (c *conn) EnsureObject(objectId string) *Object {
	c.lock.RLock()
	defer c.lock.RUnlock()
	var o *Object
	if _, ok := c.objects[objectId]; ok {
		o = c.objects[objectId]
	} else {
		o = NewObject(c.nc, objectId)
		c.objects[objectId] = o
	}
	o.Subscribe()
	return o
}
